Navigator.share() rejects with wrong exception when called multiple times

marcoscaceres opened the pull request:
#### 7654da12692d947efe23243092725dc6abc96436
<pre>
Navigator.share() rejects with wrong exception when called multiple times
<a href="https://bugs.webkit.org/show_bug.cgi?id=243652">https://bugs.webkit.org/show_bug.cgi?id=243652</a>

Reviewed by Chris Dumez.

Unfortunately, `run-webkit-tests` doesn&apos;t cause `isControlledByAutomation()`
to return `true`, so we can&apos;t test this directly. Instead, it needs to be run
with directly with web driver (as is done on WPT).

* LayoutTests/platform/ios-simulator-wk2/imported/w3c/web-platform-tests/web-share/share-sharePromise-internal-slot.https-expected.txt: Renamed from LayoutTests/platform/ios-simulator-wk2/share-sharePromise-internal-slot.https-expected.txt.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/web-share/share-sharePromise-internal-slot.https-expected.txt: Renamed from LayoutTests/platform/ios-wk2/share-sharePromise-internal-slot.https-expected.txt.
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::share):
(WebCore::Navigator::showShareData):

Canonical link: <a href="https://commits.webkit.org/253419@main">https://commits.webkit.org/253419@main</a>
</pre>

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -209,12 +209,16 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
+
     if (frame->page()->isControlledByAutomation()) {
-        promise->resolve();
+        RunLoop::main().dispatch([promise = WTFMove(promise), this] {
```
There is no guarantee that `this` is still alive by the time the lambda runs (asynchronously). You should capture `weakThis = WeakPtr { *this }` instead.

marcoscaceres replied:
ah, good point.

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -209,12 +209,16 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
     if (frame->page()->isControlledByAutomation()) {
-        promise->resolve();
+        RunLoop::main().dispatch([promise = WTFMove(promise), this] {
+            m_hasPendingShare = false;
```
Then do:
```
if (weakThis)
    weakThis->m_hasPendingShare = false;
```

cdumez commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -209,12 +209,16 @@ void Navigator::showShareData(ExceptionOr<ShareDataWithParsedURL&> readData, Ref
-    m_hasPendingShare = true;
     auto shareData = readData.returnValue();
     
     frame->page()->chrome().showShareSheet(shareData, [promise = WTFMove(promise), this] (bool completed) {
```
Please fix this one too.

cdumez reviewed (approved):
LGTM

marcoscaceres committed:
Navigator.share() rejects with wrong exception when called multiple times
https://bugs.webkit.org/show_bug.cgi?id=243652

Reviewed by Chris Dumez.

Unfortunately, `run-webkit-tests` doesn't cause `isControlledByAutomation()`
to return `true`, so we can't test this directly. Instead, it needs to be run
with directly with web driver (as is done on WPT).

* LayoutTests/platform/ios-simulator-wk2/imported/w3c/web-platform-tests/web-share/share-sharePromise-internal-slot.https-expected.txt: Renamed from LayoutTests/platform/ios-simulator-wk2/share-sharePromise-internal-slot.https-expected.txt.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/web-share/share-sharePromise-internal-slot.https-expected.txt: Renamed from LayoutTests/platform/ios-wk2/share-sharePromise-internal-slot.https-expected.txt.
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::share):
(WebCore::Navigator::showShareData):

Canonical link: https://commits.webkit.org/253419@main
